spices/SPICE-0004-HTTP-Proxy-Support.adoc

```diff
@@ -0,0 +1,180 @@
+// suppress inspection "HttpUrlsUsage" for whole file
+= HTTP Proxy Support
+
+* Proposal: link:./SPICE-0004-HTTP-Proxy-Support.adoc[SPICE-0004]
+* Author: link:https://github.com/bioball[Dan Chao]
+* Status: TBD
+* Implemented in: TBD
+* Category: Language, Standard Library
+
+== Introduction
+
+This SPICE describes support for HTTP proxying.
+
+== Motivation
+
+Some environments do not allow direct connections to the internet, instead requiring that connections go through an intermediary proxy.
+
+To enable Pkl in these situations, this SPICE provides a way to redirect outbound requests through said proxies.
+
+== Proposed Solution
+
+Pkl will support HTTP proxies.
+Proxy settings will be added, and this will control how Pkl makes HTTP requests.
+
+== Detailed design
+
+=== Configuring proxies
+
+A proxy can be set via the following, in order of precedence (least to greatest):
+
+1. `~/.pkl/settings` file
+2. `PklProject` file (in `evaluatorSettings`)
+3. CLI Flags/pkl-gradle options/`HttpClient#Builder`/server mode message passing API
+
+==== Environment Variables
+
+Proxies will _not_ be configured via environment variables.
+This maintains consistency with the rest of Pkl; it does not support using environment variables for any other configuration.
+Additionally, while other languages might use environment variables, they have slightly different meaning from one language to another (see link:https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy[]]).
+Not using environment variables avoids a possble problem where the same environment variable influences both Pkl and another runtime, but in different ways.
+
+==== Settings file
+
+A new `proxy` property is added to the global settings file.
+
+[source,pkl]
+----
+module pkl.settings
+
+import "pkl:Proxy"
+
+/// HTTP proxy settings.
+@Since { version = "0.26.0" }
+proxy: Proxy?
+----
+
+==== `PklProject`
```

Comment:
What’s the use case for this?

Reply:
I am deploying into an environment that needs proxying. Instead of setting it up on my target machine, or setting it up via a script, I can just set it in my PklProject file.

Reply:
Do you have evidence that this is useful? Is it practical to only have per-project proxy settings?

Reply:
I think the rationale here is that `settings.pkl` is a user specific configuration, which only leaves `PklProject` and CLI/Evaluator options left for shared configs. Configuring your proxy every time via CLI doesn't seem ideal, so adding this settings to `PklProject` allows for a project-wide proxy configuration.

Reply:
> Do you have evidence that this is useful? Is it practical to only have per-project proxy settings?

Yeah, this is something that was brought up as a desired use-case from some of our users.